### DISCUSSION TITLE
Change default options for Annotations

### DISCUSSION THREAD

**Discussion by PatrickZimmerer** [stephanrauh/ngx-extended-pdf-viewer#1927] [Q&A]
Is there a way to change the default options for Annotations when using the "FreeTextEditor" and the "DrawEditor"? I tried manipulating the input fields with plain JS and the values on the slider changed but as soon as I click the PDF it gets reset to the editors' default values.

**Comment by stephanrauh** [comment]
Not yet, but I agree there should be options to define the default color / font size / opacity etc. I just haven't been able to implement that yet. Also, I wonder how to call these attributes. I'm afraid we need quite a few attributes, so it might be a good idea to spend a bit time on inventing a good naming pattern.

**Reply by PatrickZimmerer** [reply]
Do you see a way to change these default values by using plain JS? 
Changing the values from the input fields ("FreeTextEditor" / "DrawEditor") works until I click into the pdf where the annotation will be created, then it gets reset to the pdf viewer's default values, or is there an event that's emitted when I click into the PDF which I can listen to and change these values on this event.

I'd call them something like this:
defaultFreeTextFontSize
defaultFreeTextColor

defaultDrawColor
defaultDrawOpacity
defaultDrawSize

maybe you could spare the "default" words to shrink the words a bit even though I like a longer name that tells me what it's doing right away.

**Reply by stephanrauh** [reply]
At the moment, I don't have a clue. But here's the source code: https://github.com/stephanrauh/pdf.js/tree/bleeding-edge/src/display/editor. It's not much, I guess (or hope) you can read it easily.

**Reply by stephanrauh** [reply]
I've opened a ticket for you: #1941

**Comment by Tweniee** [comment]
So as far as i can understand we just need to inititalize a default color and the user can change the color afterwards not restricting the user.

**Reply by stephanrauh** [reply]
Both, I guess. Some developers need a default color, while other want to restrict the choice of color. Come to think of it, there are probably projects out there that want to enforce a certain color palette. I'm afraid #1941 opens a Pandora's box of feature requests. :)

**Comment by stephanrauh** [comment]
I've found the solution. The code snippet below is pure JavaScript. You can enter it in the browser and see the effect in the UI.

```ts
const editorType = 11;
const fontSize = 42;
PDFViewerApplication.eventBus.dispatch('switchannotationeditorparams', {type: editorType, value:fontSize})
PDFViewerApplication.eventBus.dispatch('annotationeditorparamschanged', {details: [[editorType, fontSize]]});
```

Note the double brackets in the second line. It's an array of arrays, so you can modify multiple parameters at once.

These are the editor parameter types you can use: 

```ts
const AnnotationEditorParamsType = {
  RESIZE: 1,
  CREATE: 2,
  FREETEXT_SIZE: 11,
  FREETEXT_COLOR: 12,
  FREETEXT_OPACITY: 13,
  INK_COLOR: 21,
  INK_THICKNESS: 22,
  INK_OPACITY: 23,
  HIGHLIGHT_COLOR: 31,
  HIGHLIGHT_DEFAULT_COLOR: 32,
  HIGHLIGHT_THICKNESS: 33,
  HIGHLIGHT_FREE: 34,
  HIGHLIGHT_SHOW_ALL: 35,
};
```
I'll add this feature to the `NgxExtenderPdfViewerService`, so developers can find it more easily.

**Reply by stephanrauh** [reply]
I've opened a feature ticket to follow the progress (#2208).
